CFEP 03 - Providing manually uploaded builds on the conda-forge anaconda channel (#5)
conda-forge/cfep: cfep-03.md

<table>
<tr><td> Title </td><td> Providing manually uploaded builds on the conda-forge anaconda channel</td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Hadrien Mary &lt;hadrien.mary@gmail.com&gt;</td></tr>
<tr><td> Created </td><td> Sep 27, 2016</td></tr>
<tr><td> Updated </td><td> Oct 17, 2019</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Some packages, for example Qt and gcc, are very hard to build on the currently used CIs (CircleCI, Travis, AppVeyor, and AzurePipelines) due to time limitation.

> While the is a small set of packages we need a protocol to upload builds on conda-forge anaconda channel that are built off-CIs (local builds).

Reviewer: ```suggestion
While this is a small set of packages, we need a protocol to upload builds on conda-forge anaconda channel that are built off-CIs (local builds).
```

Reviewer 2: For some odd reason GH is giving me a `You do not have permission to push to this repository.` when I try to merge this :-/

## Specification

The protocol we propose is:

- The package needs a feedstock will be maintained via PRs as any other package;

> - When a PR is approved a core member will build the package locally, following the [conda-forge docs](https://conda-forge.org/docs/maintainer/updating_pkgs.html#testing-changes-locally) for local builds. Always be sure to check the docs for the latest best practices.

Reviewer: Why a core member? We've had regular members building packages like `qt`

Reviewer 2: > Why a core member?

Ideally to increase the trust in the build.

> We've had regular members building packages like `qt`

Really? Who? As far as I recall only core members did it.

Reviewer: mingwandroid, gillins.

Reviewer 2: @mingwandroid and @gillins are both kind of an unofficial core and so deeply involved with qt that they are exceptions and not the rule.

Opening for non-core folks to build something and then pressure us to upload may be a source of issues IMO and I would not like to open that can of worms.

Reviewer 2: Let's say someone approved by core.

> - When a PR is approved a core member will build the package locally, following the [conda-forge docs](https://conda-forge.org/docs/maintainer/updating_pkgs.html?highlight=local build#testing-changes-locally) for local builds. Always be sure to check the docs for the latest best practices.

Reviewer 3: This link looks broken?

Reviewer 2: We don't need the `highlight=local` but it is not broken to me. I get directed to the right page.

Reviewer 3: Sorry, I meant that the highlight=local was breaking the md rendering.

Reviewer 2: Ah. Probably! Fixed that anyways b/c it was copy-n-pasta.

> - Core members can then test and upload the packages. Not that manual uploads have the uploader username, serving as a maker for manual uploads and a tracker for asking questions regarding the build environment used.

Reviewer: Can you also add that a log is posted?

Reviewer 2: How about a new item like:

    - The build logs should be saved and posted in the feedstock PR.

Reviewer 2: BTW, how does one do something like `python build-locally.py 2>&1 | tee log` on Windows? I'd like to add that to the docs.

<table>
<tr><td> Title </td><td> Providing manually uploaded builds on the conda-forge anaconda channel</td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Hadrien Mary &lt;hadrien.mary@gmail.com&gt;</td></tr>
<tr><td> Created </td><td> Sep 27, 2016</td></tr>
<tr><td> Updated </td><td> Oct 17, 2019</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Some packages, for example Qt and gcc, are very hard to build on the currently used CIs (CircleCI, Travis, AppVeyor, and AzurePipelines) due to time limitation.

While this is a small set of packages we need a protocol to upload builds on conda-forge anaconda channel that are built off-CIs (local builds).

## Specification

The protocol we propose is:

> - The package needs a feedstock will be maintained via PRs as any other package;

Reviewer: ```suggestion
- The package needs a feedstock and will be maintained via PRs as any other package;
```

<table>
<tr><td> Title </td><td> Providing manually uploaded builds on the conda-forge anaconda channel</td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Hadrien Mary &lt;hadrien.mary@gmail.com&gt;</td></tr>
<tr><td> Created </td><td> Sep 27, 2016</td></tr>
<tr><td> Updated </td><td> Oct 17, 2019</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

Some packages, for example Qt and gcc, are very hard to build on the currently used CIs (CircleCI, Travis, AppVeyor, and AzurePipelines) due to time limitation.

> While this is a small set of packages we need a protocol to upload builds on conda-forge anaconda channel that are built off-CIs (local builds).

Reviewer 4: comma after packages; "anaconda" seems redundant after "conda-forge" (or maybe missing "and"/"or"?)

Reviewer 2: it is conda-forge anaconda.org channel but I guess we can simplify that by saying just conda-forge's channel at this point.

Note that this is a "last resort" option and we should always try to build on CIs first!

## Specification

The protocol we propose is:

- The package needs a feedstock and will be maintained via PRs as any other package;

> - If the CI time limits are the only impediment for the build one should ping the core team in the feedstock and a local build;

Reviewer: ```suggestion
- If the CI resources (which includes time, memory, disk-space and special hardware) are the only impediments for the build, one should ping the core team in the feedstock and a local build;
```

Reviewer 2: Funny. I cannot merge that from here"

![Screenshot from 2019-11-14 14-33-25](https://user-images.githubusercontent.com/950575/68881412-dc389b80-06eb-11ea-8e50-ff40fadba15b.png)

> - If the CI resources (which includes time, memory, disk-space and special hardware) are the only impediments for the build, one should ping the core team in the feedstock and a local build;

Reviewer 5: > and a local build

"prepare" or "setup up" or which verb is missing here?

Reviewer 2: Let's go with prepare.

> - When a PR is approved, someone "blessed" by a core member will build the package locally, following the [conda-forge docs](https://conda-forge.org/docs/maintainer/updating_pkgs.html#testing-changes-locally) for local builds. Always be sure to check the docs for the latest best practices.

Reviewer 3: ```suggestion
- When a PR is approved, someone authorized by a core member will build the package locally, following the [conda-forge docs](https://conda-forge.org/docs/maintainer/updating_pkgs.html#testing-changes-locally) for local builds. Always be sure to check the docs for the latest best practices.
```

> - When a PR is approved, someone authorized by a core member will build the package locally, following the [conda-forge docs](https://conda-forge.org/docs/maintainer/updating_pkgs.html#testing-changes-locally) for local builds. Always be sure to check the docs for the latest best practices.

Reviewer 3: Any thoughts on how this authorization would work? Is it per PR, a time limit, a list of people we keep somewhere? (A github team?)

Reviewer 2: Only a core member can upload it, so the authorization here does not mean much. This was added only so others "blessed" by a core member can build it.

Reviewer 3: Ah, ok. Thank you!

- The build logs should be saved and posted in the feedstock PR.

> - Core members can then test and upload the packages. Note that manual uploads have the uploader username, serving as a maker for manual uploads and a tracker for asking questions regarding the build environment used. The build logs should be saved and posted in the feedstock PR. (One easy way to achieve that on Linux is to use `python build-locally.py 2>&1 | tee log.txt` to build).

Reviewer 4: "maker" -> "marker"?

## Motivation

Main motivation is to be able to provide hard to compile/long builds on conda-forge.

## Alternatives

> A long term solution would be dedicated machines that we could setup via a CI service, like AzurePipelines.

Reviewer 4: FWIW, I've opened conda-forge/conda-forge.github.io#924 to explore this some more with the azure folks.

Reviewer 4: Someone from azure is looking into it; note that the discussion moved to conda-forge/conda-forge.github.io#902.

<table>
<tr><td> Title </td><td> Providing manually uploaded builds on the conda-forge anaconda channel</td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Hadrien Mary &lt;hadrien.mary@gmail.com&gt;</td></tr>
<tr><td> Created </td><td> Sept 27, 2016</td></tr>
<tr><td> Updated </td><td> Sept 27, 2016</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

> Some packages are very hard to build on the currently used CI (circleci, Travis and AppVeyor). It's most of the time because of time limitations in the CI website and also because of the complexity of the build.

Reviewer 6: nit: `circleci` -> `CircleCI`

Reviewer 2: Fixed.

An example is the Qt5 build that seems to be very hard to build (it tooks almost a year for Continuum folks to compile it for all platforms). See the Qt5 recipe build discussion ticket here https://github.com/conda-forge/qt-feedstock/issues/5.

That being said it's excepted that a small amount of packages will be very hard to automatically build for various reasons. 

> Obviously we all want to keep this amount of packages small and have a maximun of automatically build packages on conda-forge.

Reviewer 6: typo: `this amount` -> `the amount`

nit: `maximun` -> `maximum` or just drop this part.

Reviewer 2: Fixed.

But a system/protocol to manually upload builds on conda-froge anaconda channel is necessary to circumvent some situations.

## Specification

A protocol to manually upload builds could look something like that :

> - Manually build the different packages for all targeted platforms.

Reviewer 6: Could we outline a bit more the steps that go into this? For instance, on Linux we can use our Docker container to build packages. On Windows, maybe we could use this ( https://github.com/ContinuumIO/vagrant-images/pull/1 ). On OS X, I'm not sure what we can do ATM, but it would be nice if we could give that some thought too.

Reviewer 2: In the new CEP-03 this points to the docs that are more dynamic and avoid duplication.

> - Test for each platform on different version (multiple Linux distributions, Windows 7/10, etc).

Reviewer 6: Could we specify a set list of acceptable Linux distro that qualify as sufficient?

Reviewer 7: I'd say CentOS 6

- Someone from the @conda-forge/core (having the permissions to upload on the conda channel) use conda command line to upload the package.

> - Build is available in the conda-forge anaconda channel.

Reviewer 6: So one thought on this. Maybe so we can make it clear if something was manually built or not, would be to put this under a different label other than `main` like `manual`. We can still include this as a source at conda-forge, but it would be something one would need to add explicitly locally.

Reviewer 2: The information is on the uploader username. No need for a new label.

**Open Ideas/Questions**

- Could we use CI to test manually uploaded packages ? How ?

> - How the package is transfered from the user to the @conda-forge/core person before the upload ?

Reviewer 6: My gut feeling is a `conda-forge/core` person would do the build with a standard VM. So they could do the upload too.

Reviewer 7: :+1:

Reviewer 2: That is in the new text.

## Motivation

Main motivation is to be able to provide the widest range of possible packages in conda-forge. This proposal will be really usefull to upload very hard to build packages.

## Alternatives

A CI infrastructure that build them all !

## FAQ

## Copyright

> This document has been placed in the public domain.

Reviewer 6: Please change to `This document is CC0 1.0 Universal.` Not everywhere has Public Domain, but this is the next best thing. IANAL

<table>
<tr><td> Title </td><td> Providing manually uploaded builds on the conda-forge anaconda channel</td>
<tr><td> Status </td><td> Proposed </td></tr>
<tr><td> Author(s) </td><td> Hadrien Mary &lt;hadrien.mary@gmail.com&gt;</td></tr>
<tr><td> Created </td><td> Sept 27, 2016</td></tr>
<tr><td> Updated </td><td> Sept 27, 2016</td></tr>
<tr><td> Discussion </td><td> NA </td></tr>
<tr><td> Implementation </td><td> NA </td></tr>
</table>

## Abstract

> Some packages are very hard to build on the currently used CI (CircleCI, Travis and AppVeyor). It's most of the time because of time limitations in the CI website and also because of the complexity of the build.

Reviewer 8: Could you clarify what you mean by complexity of the build? If it wasn't for the time limits, why would a manual/local build be better than an CI build?

Author: On this maybe @ccordoba12 could describe more precisely why does it take so long to get the Qt build.

Naively I would say that the huge amount of C/C++ dependencies in a project make the process to build it harder than without dependencies. And if you add the fact that you have to wait for the CI to end before being able to do a modification and try to compile again, it makes the whole process discouraging.

Reviewer 7: Qt 5 is a beast! Besides its standard libraries, it also compiles Chromium to create bindings for it :-)

Reviewer 2: >  it also compiles Chromium to create bindings for it :-)

:scream:

Reviewer 8: I have no doubt that Qt5 is hard to build! I guess what I meant is more that imagine there were no time constraints on CI builds - would you sill want to do a Qt5 build locally? My suspicion is that we'd still want to do it on the CI - which means that build time is the _only_ acceptable reason for doing builds locally instead? (not the complexity of the build per se - you could have a build that is complex to set up but runs in 15 minutes?)

Author: I agree with you. I think if a CI we infinite build time was available we could avoid local builds.

Reviewer 7: There could be space limitations too. Qt 5 takes about 2.5 gigs to build and 1 more gig for installation (but that could change between OSes :-)

Reviewer 2: @astrofrog that is an important question and the text was not clear. The new text covers only recipes that we can "just download and build." Excluding things that are hard to create recipes but easy to build locally. The reason is to avoid hacks that may obscure the builds and ensure transparency of what will be built locally.

TL;DR this cfep covers only CI build time limitation.

> An example is the Qt5 build that seems to be very hard to build (it tooks almost a year for Continuum folks to compile it for all platforms). See the Qt5 recipe build discussion ticket here https://github.com/conda-forge/qt-feedstock/issues/5.

Reviewer 8: I might be wrong, but I don't think the complexity issue would be solved by manual uploads? In other words, is there really any reason other than build time why we need manual uploads?

Author: By building manually you can fix the build more quickly but I agree that if you run locally the same distrib/OS as the CI you should be able to backpropagate your "working" build recipe to the CI.

> An example is the Qt5 build that seems to be very hard to build (it tooks almost a year for Continuum folks to compile it for all platforms). See the Qt5 recipe build discussion ticket here https://github.com/conda-forge/qt-feedstock/issues/5.

Reviewer 9: `it tooks` > `it took`

> That being said it's excepted that a small amount of packages will be very hard to automatically build for various reasons. 

Reviewer 8: excepted -> expected

Obviously we all want to keep the amount of packages small and have a maximum of automatically build packages on conda-forge.

But a system/protocol to manually upload builds on conda-froge anaconda channel is necessary to circumvent some situations.

## Specification

A protocol to manually upload builds could look something like that :

- The user provides the recipe as a PR.

- The build fails : someone from @conda-forge/core decides to "enable" the manually uploaded builds system.

- People from @conda-forge/core are designated to build, test and upload the package.

For Linux, the `scripts/run_docker_build.sh` should be used (builds are placed in `build_artefacts/`. Building with this script will provide consistency for these packages because they will be build against the same Linux distribution.

> On Windows, we could use this : ContinuumIO/vagrant-images#1. (Need help here I have no expertise with Windows)

Reviewer 10: We would likely need to ensure we have our own vagrant image with all the correct compilers set up - but certainly vagrant is our next closest thing to docker.

Reviewer 2: Indeed. @mariusvniekerk created a VM to build `scipy` and we should document the steps he took to achieve that. However, that should be in the docs and not in the cfep.

On OS X, same comment as before I don't know/use OS X.

We also need a way to differentiate manually uploaded packages from "main" packages. Maybe using different build label `main` and `manual`.

- Test for each platform and different versions the package. The test should be the one already written in the `meta.yaml` file.

For Linux, test at least Debian based distribution and Fedora/CentOS. (Any suggestions for more Linux distribution ?).

For windows, test at least Windows 7 and Windows 10. (do we need Windows 7 ?)

For OS X, we should test against version 10.9 and 10.11.

- Upload the packages.

> Note about security : @conda-forge/core people would need to be extremly trusted beccause they are going to build the packages and so could potentially introduce security breach in binary files. An auditing process should be considered (I don't see one at the moment except from trusting people). Also adding the conda recipe to the package is a good practice I think. (Now when you think about it, the situation is not very different from PyPi wheels that contains binary files.)

Reviewer 11: *because

Reviewer 6: Also *extremely .

Do you have LibreOffice or something @hadim ? I bet a simple run through of spell check will catch these nits. Just thinking so as to save you and reviewers time.

Author: Sorry guys but I won't be able to put more time on this before January. If someone want to take the lead on this PR that would be nice or I can do it later in January.

**Open Ideas/Questions**

> - Could we use CI to test manually uploaded packages ? How ?

Reviewer 8: What about uploading them manually to a `manual` label then opening a PR to indicate that a manual build has been added, which will then trigger the testing phase to be run on CI, and if it passes the build can be copied over to the main label?

Author: By doing that it would impossible to differentiate a manually build package after it has been accepted from a CI build package. Am I correct ?

Reviewer 11: Yes, but I think testing the package ensures that the tests executed when building the package were not indicating a false-positive due to some local configuration, or if the person building it passed `--no-test` for some reason. 

Testing the package in an isolated and trusted environment out-weights that small disadvantage, IMHO.

Reviewer 2: We have an identifier, that uploader username.

- The package needs a feedstock and will be maintained via PRs as any other package;

> - If the CI resources (which includes time, memory, disk-space and special hardware) are the only impediments for the build, one should ping the core team in the feedstock and a setup up local build;

Reviewer 5: "...up up" and away ;P

A protocol to manually upload builds could look something like that :

- The user provides the recipe as a PR.

- The build fails : someone from @conda-forge/core decides to "enable" the manually uploaded builds system.

- People from @conda-forge/core are designated to build, test and upload the package.

> For Linux, the `scripts/run_docker_build.sh' should be used (does it needs to be modified ?). Building with this script will provide consistency for these packages because they will be build against the same Linux distribution.

Reviewer 6: > scripts/run_docker_build.sh' should be used (does it needs to be modified ?)

I don't believe so. This should build and place everything in a directory in the repo in a directory called `build_artefacts`.

nit: `scripts/run_docker_build.sh' -> `scripts/run_docker_build.sh` ( last character should be ` )

Reviewer 2: The new text points to the docs to avoid duplication.

On Windows, we could use this : ContinuumIO/vagrant-images#1. (Need help here I have no expertise with Windows)

> On OS X, same comment as before I don't know/use OS X.

Reviewer 6: There are some images being prepped, but there is a question as to what licensing constraints apply.

Reviewer 2: We need to add that to the docs and not in the cfep.

We also need a way to differentiate manually uploaded packages from "main" packages. Maybe using different build label `main` and `manual`.

- Test for each platform and different versions the package. The test should be the one already written in the `meta.yaml` file.

For Linux, test at least Debian based distribution and Fedora/CentOS. (Any suggestions for more Linux distribution ?).

For windows, test at least Windows 7 and Windows 10. (do we need Windows 7 ?)

> For OS X, without expertise it's hard for me to tell which version should be tested.

Reviewer 6: Currently we support 10.9 a.k.a. Mavericks. Maybe testing on 10.11 a.k.a. El Capitan.

Reviewer 2: No need to specify this in the cfep. It should be in the docs.

- Upload the packages.

> Note about security : @conda-forge/core people would need to be extremly trusted beccause they are going to build the packages and so could potentially introduce security breach in binary files.

Reviewer 6: Honest question. Is there any auditing process that you think would be valuable here?

Note: Anaconda.org does keep a log of all actions that take place on a channel.

Reviewer 2: That is a question for the Anaconda.org team. Which BTW is being redesigned.
